Implementation of MULTILEADER and MULTILEADERSTYLE

mme1950 opened the pull request.

DomCR commented on ACadSharp/IO/DWG/DwgStreamReaders/DwgObjectReader.cs:
```diff
@@ -1585,11 +1593,7 @@ private CadTemplate readPolyline3D()
-			{
-				pline.Flags |= PolylineFlags.ClosedPolylineOrClosedPolygonMeshInM;
-			}
+			bool closed = (this._objectReader.ReadByte() & 1U) > 0U;
```
Changing the method readPolyline3D() seems to be outside of the scope for this task please revert the changes, unless there is a major reason to change this method.

DomCR commented on ACadSharp/IO/DWG/DwgStreamReaders/DwgObjectReader.cs:
```diff
@@ -2797,6 +2801,574 @@ private CadTemplate readLeader()
+		}
+
+
+		private void analyse02(int count = 20) {
```
If I'm not mistaken the methods `analyse02` and `analyse03` are implemented to explore the stream values, please do not include them in the final PR.

DomCR commented on ACadSharp/IO/DWG/DwgStreamReaders/DwgObjectReader.cs:
```diff
@@ -4534,8 +5106,8 @@ private CadTemplate readVisualStyle()
-			var objValues = DwgStreamReaderBase.Explore(_objectReader);
-			var textValues = DwgStreamReaderBase.Explore(_textReader);
+			//var objValues = DwgStreamReaderBase.Explore(_objectReader);
+			//var textValues = DwgStreamReaderBase.Explore(_textReader);
```
`readVisualStyle()` is outside the scope of this task, please revert the changes.

DomCR commented on ACadSharp/IO/DXF/DxfStreamReader/DxfObjectsSectionReader.cs:
```diff
@@ -59,8 +59,8 @@ private CadTemplate readObject()
-				//case DxfFileToken.ObjectSortEntsTable:
-				//return this.readSortentsTable();
+				case DxfFileToken.ObjectSortEntsTable:
+					return this.readSortentsTable();
```
currently there is an error with the `readSortentsTable()` don't uncomment the method.

DomCR commented on ACadSharp/IO/DXF/DxfStreamReader/DxfSectionReaderBase.cs:
```diff
@@ -585,12 +585,6 @@ private bool readLwPolyline(CadEntityTemplate template, DxfMap map, string subcl
-					{
-						last.CurveTangent = this._reader.ValueAsDouble;
-					}
-					return true;
```
Why has this been removed?

mme1950 replied:
As before. We did not make any changes to this class.

DomCR commented on ACadSharp/IO/DXF/DxfStreamWriter/DxfHeaderSectionWriter.cs:
```diff
@@ -10,11 +9,8 @@ internal class DxfHeaderSectionWriter : DxfSectionWriterBase
 
-		public DxfWriterOptions Options { get; }
-
-		public DxfHeaderSectionWriter(IDxfStreamWriter writer, CadDocument document, CadObjectHolder holder, DxfWriterOptions options) : base(writer, document, holder)
```
Options should not be removed from this class.

mme1950 replied:
As before. We did not make any changes to this class.

DomCR commented on ACadSharp/IO/DXF/DxfStreamWriter/DxfHeaderSectionWriter.cs:
```diff
@@ -23,9 +19,6 @@ protected override void writeSection()
 
 			foreach (KeyValuePair<string, CadSystemVariable> item in map)
 			{
-				if (!this.Options.WriteAllHeaderVariables && !this.Options.HeaderVariables.Contains(item.Key))
```
This if is needed to avoid errors in the writer, please revert the changes.

DomCR commented on ACadSharp/IO/DXF/DxfStreamWriter/DxfObjectsSectionWriter.cs:
```diff
@@ -37,7 +37,7 @@ protected void writeObject<T>(T co)
 					break;
 				case SortEntitiesTable sortensTable:
-					//this.writeSortentsTable(sortensTable);
+					this.writeSortentsTable(sortensTable);
```
Same as the reader `writeSortentsTable(sortensTable)` has some errors, please do not uncomment this line.

DomCR commented on ACadSharp/IO/Templates/CadInsertTemplate.cs:
```diff
@@ -34,12 +34,7 @@ public override void Build(CadDocumentBuilder builder)
-				insert.Block = block;
-			}
-			else if (!string.IsNullOrEmpty(this.BlockName) && builder.TryGetTableEntry(this.BlockName, out block))
+			if (builder.TryGetCadObject(this.BlockHeaderHandle, out BlockRecord block))
```
This is an old if, please revert to the new one.

DomCR commented:
First of all, great work!! this looks great!! to continue with the development I would like to add a few notes:

- Add test to check that the `MLEADER` and `MLEADERSTYLE` are read without errors, the current test files only contain `MLEADERSTYLE`, you should add a file in each version (if possible) containing `MLEADER` in it.
- The `DxfReader` and `DxfWriter` haven't been updated, do you plan to do it in this PR or keep it separately? 
- When updating the branch be careful for which are the last changes in master, some of my comments were related to this issue.

Overall looks good, I will continue to review the PR while you work on it and continue to give feedback to you.

Thanks for the contribution.

mme1950 commented on ACadSharp/DxfPropertyBase.cs:
```diff
@@ -153,10 +153,6 @@ public void SetValue<TCadObject>(int code, TCadObject obj, object value)
 			{
 				this._property.SetValue(obj, Enum.ToObject(this._property.PropertyType, value));
 			}
-			else if (this._property.PropertyType.IsEquivalentTo(typeof(ushort)))
```
I am afraid, we did something wrong. We did not change this file. We have to check how this happened.

mme1950 commented on ACadSharp/Entities/MultiLeader.cs:
```diff
@@ -0,0 +1,319 @@
+
+
+		/// <summary>
+		/// TODO Do we need this class to represent a list of arrow heads
```
I pulled the TODO out of the summary.
I think we do not need the ArrowheadAssociation class.
The Arrowheads property should have the type IList<Block> (or whatever type an arrowhead is)

mme1950 commented on ACadSharp/Entities/MultiLeader.cs:
```diff
@@ -0,0 +1,319 @@
+		/// <summary>
+		/// Text attachment direction for MText contents
+		/// 	0 = Horizontal
+		/// 	1 = Vertical
```
It is copied from the documentation, just to be sure that the created enum types are correct.
I moved itnto the value tag, OK?

DomCR replied:
Value tag should only be used to add a commentary about the value restriction, if is an enumeration you don't need to specify the values, you can remove the tag.

mme1950 commented on ACadSharp/Entities/MultiLeader.cs:
```diff
@@ -0,0 +1,319 @@
+
+		public override CadObject Clone()
+		{
+			MultiLeader clone = (MultiLeader)base.Clone();
```
OK

mme1950 commented on ACadSharp/Header/CadHeader.cs:
```diff
@@ -1573,7 +1573,7 @@ public XYZ ModelSpaceYAxis
 		[CadSystemVariable("$TIMEZONE", 70)]
 		public int TimeZone { get; set; }
-
+		
```
There should not be any changes in this class.

mme1950 commented:
I am afraid, we did something wrong. Some of the changes you mentioned above have not been made by us. We have to check how this happened.

Maybe we have to redo the fork, create a branch and enter a PR.
I will return to it tomorrow.

mme1950 commented:
* Add test to check that the MLEADER and MLEADERSTYLE are read without errors, the current test files only contain MLEADERSTYLE, you should add a file in each version (if possible) containing MLEADER in it.

> OK, we must view the test project.


* The DxfReader and DxfWriter haven't been updated, do you plan to do it in this PR or keep it separately?

> We then also must add `writeMultiLeader` and `writeMultiLeaderStyle` to `DwgObjectWriter`, right?
> I would prefer do do it separately.


* When updating the branch be careful for which are the last changes in master, some of my comments were related to this issue.

> OK

DomCR commented:
I've been reviewing the PR, for some reason your PR is pointing to an old version of master and registering a lot of changes that should not be there... I don't know why is this way, I'll try to find out more about this issue.

In the meantime keep an eye to the test, I've approved the build but is failing, check what is happening, it may be caused due this unnecessary changes.

mme1950 commented:
For a full test we have to create a DWG files with "all" variants of Mutileader also for older AutoCAD versions, right?
To make the test run a subset may be sufficient.
For all variants we have to learn about block content, arrow heads and some other things we did not try yet.

DomCR commented:
> For a full test we have to create a DWG files with "all" variants of Mutileader also for older AutoCAD versions, right? To make the test run a subset may be sufficient. For all variants we have to learn about block content, arrow heads and some other things we did not try yet.

If you want to just test in a single version you have to make sure that the reader will ignore the entities for all other versions to avoid issues or bugs.

DomCR commented on ACadSharp/Objects/MultiLeaderAnnotContextClasses.cs:
```diff
@@ -0,0 +1,223 @@
+	/// Common AcDbAnnotScaleObjectContextData data (see paragraph 20.4.71).
+	/// 300 DXF: “CONTEXT_DATA{“
+	/// </summary>
+	public partial class MultiLeaderAnnotContext : CadObject
```
The documentation for this class should be revised, I see that is copied from the PDF.

DomCR commented on ACadSharp/Tables/MultiLeaderStyle.cs:
```diff
@@ -0,0 +1,312 @@
+		/// <value>
+		/// 	0 = Horizontal
+		/// 	1 = Vertical
+		/// </value>
```
The value tag for the documentation should be only used to add value restrictions like a range of values, enumeration types don't need to be specified.

DomCR commented:
It seems that some tests are failing, one of them is the `DwgReaderTest` the test is making a comparison with a json file that does not belong to the new sample, the best way to go it will be to move the sample to another folder and create an specific test for it.

Also I see that the Clone() test is failing too for the `MultiLeader` it goes in some kind of loop.

Let me know if you need any help with that.

mme1950 commented:
> It seems that some tests are failing, one of them is the `DwgReaderTest` the test is making a comparison with a json file that does not belong to the new sample, the best way to go it will be to move the sample to another folder and create an specific test for it.
> 
> Also I see that the Clone() test is failing too for the `MultiLeader` it goes in some kind of loop.
> 
> Let me know if you need any help with that.

Hi @DomCR,

I am sorry, I did not have time to work on this project the last two weeks.
I return today to analyse these problems with Nihad.

Then we should try to finish.
I would like to solve the problems regarding the version compliance by simply not supporting mutileaders form AutoCAD versions older than R2010.
I think the object model is OK now.

DomCR commented:
> > It seems that some tests are failing, one of them is the `DwgReaderTest` the test is making a comparison with a json file that does not belong to the new sample, the best way to go it will be to move the sample to another folder and create an specific test for it.
> > Also I see that the Clone() test is failing too for the `MultiLeader` it goes in some kind of loop.
> > Let me know if you need any help with that.
> 
> Hi @DomCR,
> 
> I am sorry, I did not have time to work on this project the last two weeks. I return today to analyse these problems with Nihad.
> 
> Then we should try to finish. I would like to solve the problems regarding the version compliance by simply not supporting mutileaders form AutoCAD versions older than R2010. I think the object model is OK now.

The PR and the implementation looks good, and if you want to focus on one feature at a time that's ok, even if you want to split the PR that would be OK for me too. 

Just as a general note, check the documentation comments and avoid to use the /// <value> tag for enums and keep in mind to delete the development methods.

mme1950 commented:
Hi @DomCR,

is it OK for you, to restrict ths support of multileader entities to version R2010 Plus?
Then I will change the implementation accordingly.

Here are some remarks to our implementation:

## Object Model
According to the PDF (OpenDesign_Specification_for_.dwg_files) and online documentation the MultiLeader entity  embeds an object that provides a subset ob the properties of a ```MLeaderAnnotContext``` object. The ```MLeaderAnnotContext``` includes a list of embedded LEADER objects. LEADER objects contain a list of LEDER_LINE objects.

> In the DXF file with a MULTILEADER entity we also found a complete ```MLeaderAnnotContext``` object with the common properties and properties identical to the object embedded into the MULTILEADER.

This structure is reproduced in the object model:
* The ```MultiLeader``` class has a ```Style``` property referencing a ```MultiLeaderStyle``` object.
* A ```MultiLeaderAnnotContext``` class is defined holding the subset of properties from the ```MLeaderAnnotContext```. Each ```MultiLeader``` references exactly one ```MultiLeaderAnnotContext```.
* The ```MultiLeaderAnnotContext``` has a list of ```LeaderRoot``` objects.
* A LederRoot has a list of ```LeaderLine``` objects.

Several properties of the ```MultiLeader``` class also exist in the ```MultiLeaderAnnotContext``` class. Values may be redundant or conflicting.


## Remarks
In AutoCAD a MultiLeader can have one or more leaders evolving to leader lines (LEADER_LINE). It seems there is no way to add more than one leader (LEADER), having one or more leader lines each.

Thus: There can be only one LEADER. More than one LEADER would require an individual Contents. But the MultiLeader can have has one "Contents", either a Text contents or a contents block.

The ```MultiLeader``` class has a ```LineType``` property enherited from ```Entity``` referencing a ```LineType``` object. A second property of type ```LineType``` (```LeaderLineType```) is defined in the ```MultiLeader``` class. The referenced ```LineType``` objects are different.


## Open Issues

### Version Compliance
Due to lack of documentation and because we cannot create and test DWG-files with old AutoCAD versions we can support multileaders only for R2010 Plus.

### List of Arrowheads
In the PDF (OpenDesign_Specification_for_.dwg_files) documentation and also in the various DXF documentations a list of arrowheads is mentioned. This data are obviously not present in the DWG. We assume that the documentation is wrong.


### Two Unknown Bytes in MultiLeader
Between *291 Enable Dogleg* and *41  Dogleg Length* we must call Advance(2).

mme1950 commented:
Hi @DomCR,

final changes (hopefully):
* removed all enum values from /// comments, except where only a subset is mentioned in the documentation.
* fixed formatting (opening brace on new line).
* removed analyse methods.
* `readMultiLeader` and `readMultiLeaderStyle` return **null** when version is older than R2010.

regards
Matthias

mme1950 commented:
Hi @DomCR,

please commit this PR if the code is OK for you now.

regards
Matthias

DomCR commented on ACadSharp/LeaderLinePropertOverrideFlags.cs:
```diff
@@ -0,0 +1,43 @@
+
+namespace ACadSharp {
+
+	public enum LeaderLinePropertOverrideFlags : Int32 {
```
If this is enum works as a flag you need to add the Attribute `Flags` on it.

DomCR commented on ACadSharp/Objects/MultiLeaderAnnotContext.cs:
```diff
@@ -0,0 +1,362 @@
+		public double OverallScale { get; set; }
+
+		/// <summary>
+		/// 3BD	10	Content base point
```
As mentioned in https://github.com/DomCR/ACadSharp/pull/169/files#r1360745917 please remove the types and the codes from the Summary tag.

DomCR commented on ACadSharp/Objects/MultiLeaderAnnotContextClasses.cs:
```diff
@@ -0,0 +1,210 @@
+			public LeaderRoot() { }
+
+			/// <summary>
+			/// B	290	Is content valid (ODA writes true)
```
Same as for [MultiLeaderAnnotContext.cs](https://github.com/DomCR/ACadSharp/pull/169/files#diff-f431954f2bc08d2f6071c89702472b01e443aada7ee63b093f5d00d6a82995ea)

DomCR commented on ACadSharp/IO/DWG/DwgStreamReaders/DwgObjectReader.cs:
```diff
@@ -1206,12 +1212,27 @@ private void readCommonAttData(AttributeBase att)
 
 			switch (att.AttributeType)
 			{
+				case AttributeType.SingleLine:
```
the method `readCommonAttData(AttributeBase att)` should not change

DomCR commented on ACadSharp/IO/DWG/DwgStreamReaders/DwgObjectReader.cs:
```diff
@@ -1256,7 +1269,9 @@ private CadTemplate readDocumentTable<T>(Table<T> table, CadTableTemplate<T> tem
-				//numentries handles in the file (soft owner)
+				//Handle refs H NULL(soft pointer)
+				//xdicobjhandle(hard owner)
+				//the apps(soft owner)
```
`readDocumentTable<T>` should not change

DomCR commented on ACadSharp/IO/DWG/DwgStreamReaders/DwgObjectReader.cs:
```diff
@@ -1393,7 +1408,7 @@ private void readInsertCommonData(CadInsertTemplate template)
 
 			//R2004+:
-			if (this.R2004Plus && template.HasAtts)
+			if (this.R2004Plus & template.HasAtts)
```
`readInsertCommonData(CadInsertTemplate template)` should not change

DomCR commented on ACadSharp/IO/DWG/DwgStreamReaders/DwgObjectReader.cs:
```diff
@@ -1562,15 +1577,12 @@ private CadTemplate readPolyline3D()
 			//Flags RC 70 NOT DIRECTLY THE 75. Bit-coded (76543210):
-			byte flags = this._objectReader.ReadByte();
-
+			byte num1 = this._objectReader.ReadByte();
```
`readPolyline3D()` should not change

DomCR commented on ACadSharp/IO/DWG/DwgStreamReaders/DwgObjectReader.cs:
There are a lot of changes in methods outside the scope for this file, I think is due some merge conflicts.

Please check the changes, this file should only be changes for the MULTILEADER and MULTILEADERSTYLE implementation methods

DomCR commented:
Great Job with this one!! looks really good!!

I've reviewed the PR, most of the comments are related to the documentation or fix some merging errors that you may have had during the development, I'll try to block any PRs that mess with you files until this is merge into master.

DomCR commented:
There are some test failing:

- Failed `ACadSharp.Tests.ColorTests`: this are due the merging errors, once you restore those classes you will be good.
- Failed `ACadSharp.Tests.IO.IOTests`: this test is using the `DwgWriter` and `DxfWriter` for testing, the `ACadSharp.IO.DXF.DxfSectionWriterBase.writeEntity` and `DwgObjectWriter.Entities` are throwing an exception for your new entities, please add them into the ignore list so this exception is avoided.

DomCR commented:
If you need a hand resolving this issues, please let me know and I'll commit the changes directly to this PR.

mme1950 commented on ACadSharp/IO/DWG/DwgStreamReaders/DwgMergedReader.cs:
```diff
@@ -18,38 +17,38 @@ internal class DwgMergedReader : IDwgStreamReader
 		public void Advance(int offset)
 		{
-			throw new InvalidOperationException();
+			_mainReader.Advance(offset);
```
I agree, we did not change anything else in this file.

mme1950 commented:
Hi @DomCR,

we hopefully fixed the merge conflicts and changed the comments as you requested.
However the comments are in general very poor. I will improve them later when we learnt how properties specified in AutoCAD arrive in the data.

regards
Matthias

DomCR reviewed (commented):
The PR looks good but there is still a problem with the clone for `ACadSharp.Entities.MultiLeader` I'll take a look and see if I can help

DomCR commented:
Here is a fix for the `ACadSharp.Entities.MultiLeader.Clone()` error:

```C#
		public override CadObject Clone()
		{
			MultiLeader clone = (MultiLeader)base.Clone();

			clone.ContextData = (MultiLeaderAnnotContext)this.ContextData?.Clone();

			foreach (var att in BlockAttributes)
			{
				clone.BlockAttributes.Add(att);
			}

			return clone;
		}
```

mme1950 commented:
Oh.

DomCR commented:
> Oh.

Sorry I wanted to try if I was able to push directly here and make the merge if the test passed.

DomCR commented:
Great job!!

mme1950 commented:
Hi @DomCR

So you made the change yourself and merged everythng, Thank you wery much.
We will now concentrate on our DWG-to-SVG converter an so intensively test the multileder reader.
Then we shall implement the DXF-reader and the writers.

Again thank you very much for your help and the review. It is great to contribute to such an active project.
Greetings from Berlin to Barcelona.
